### PR TITLE
Mandelbrot: Improvements and features

### PR DESCRIPTION
**Pull request by ronak69**
If demos shouldn't be bloated with features then only consider the first 2 commits.

Feedback wanted for the architecture of one window updated by another window thing. The current implementation is very hackish.

---

#### 1. Mandelbrot: Correct a mistake in the smooth coloring math
```    
    `sqrt()` was being used instead of `log()`. Because of that little
    mistake, bands of colors are rendered and not smooth gradient as
    expected with smooth coloring.
```
**Before**: 
image

**After**:
image

#### 2. Mandelbrot: Only calculate the set in paint event
```    
    The set should be calculated only when there is a paint event to keep
    the number of times the set is calculated (which is compute expensive)
    to the minimum.
```

This significantly reduces the amount of compute. Try spam zooming in and out.

#### 3. Mandelbrot: Add functionality to change colors
```    
    Colors get chaged by adding an offset to the hue of every color that is
    calculated for the set. In other words: simply rotating the color wheel
    used by the renderer.
```
image

#### 5. Mandelbrot: Add Juila Set rendering
```    
    Adds a "Show Julia Set" option in the View menu.
    
    Ctrl+Primary-Click or Ctrl+Dragging in the Mandelbrot set window will
    render the corresponding Julia set for the complex number under the
    cursor in the Julia set window.
```

#### 6. Mandelbrot: Add statusbar to the Julia set window
```    
    The statusbar shows the complex number `c` of the rendered Julia set.
```
image

### CODE REVIEW

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 1]
```diff
@@ -118,8 +121,12 @@ class MandelbrotSet {
     void calculate_pixel(int px, int py, int max_iterations)
     {
         auto iterations = mandelbrot(px, py, max_iterations);
-        auto whole_iterations = floor(iterations);
-        auto partial_iterations = fmod(iterations, 1);
+        if (iterations == max_iterations) {
+            m_bitmap->set_pixel(px, py, Color::Black);
+            return;
+        }
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:153]: Did you implement that because the edge case was wrong, or because this is a fast path?
Except if this is closely related to the changes in `mandelbrot`, this should be in its own commit.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 2]
```diff
@@ -118,8 +121,12 @@ class MandelbrotSet {
...
+        auto whole_iterations = static_cast<int>(iterations);
+        auto partial_iterations = iterations - whole_iterations;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:155]: Why did you change these two lines?

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 3]
```diff
@@ -106,8 +106,11 @@ class MandelbrotSet {
         if (iteration == max_iterations)
             return iteration;
 
-        auto lz = sqrt(x * x + y * y) / 2;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: On the commit message:
I would suggest you to read the Contribution guidelines again, but in a nutshell, use the commit title to describe what you've done and not "Fix/Correct xxx". So prefer something like `Use log(x) instead of sqrt(x) to smooth color gradients`.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 4]
```diff
@@ -23,6 +24,8 @@
 #include <LibMain/Main.h>
 #include <unistd.h>
 
+double g_color_hue_offset = 0;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:36]: Please avoid global variable, this should be a member of `MandelbrotSet`

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 5]
```diff
@@ -473,10 +479,29 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         },
         window);
 
+    double last_saved_color_hue = 0;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:608]: Instead of having a variable outside the lambda, you can save the value just before creating the `ColorPicker` and restore it on cancellation.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 6]
```diff
@@ -185,8 +221,8 @@ class MandelbrotSet {
     {
         auto y_mid = m_y_start + (m_y_end - m_y_start) / 2;
         auto aspect_corrected_y_length = (m_x_end - m_x_start) * m_bitmap->height() / m_bitmap->width();
-        m_y_start = y_mid - aspect_corrected_y_length / 2;
-        m_y_end = y_mid + aspect_corrected_y_length / 2;
+        m_y_start = y_mid + aspect_corrected_y_length / 2;
+        m_y_end = y_mid - aspect_corrected_y_length / 2;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:227]: This looks like it's only caused by you inverting the sign of `y_start` and `y_end`

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 7]
```diff
@@ -24,11 +24,19 @@
 #include <LibMain/Main.h>
 #include <unistd.h>
 
+enum class SetType {
+    Mandelbrot,
+    Julia,
+};
+
+double g_c_real = 0.3;
+double g_c_imag = 0.01;
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:35]: Same here, no global variable please

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 8]
```diff
@@ -166,14 +190,26 @@ class MandelbrotSet {
...
+        return pixel_x * (m_x_end - m_x_start) / m_bitmap->width() + m_x_start;
+    }
+
+    double pixel_to_imag(int pixel_y) const
+    {
+        return pixel_y * (m_y_end - m_y_start) / m_bitmap->height() + m_y_start;
+    }
+
+    SetType m_set_type;
+
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:206]: All of that should be `private`. (That can be done once you put `g_c*` as a member of the class)

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 9]
```diff
@@ -366,6 +419,13 @@ void Mandelbrot::mousedown_event(GUI::MouseEvent& event)
 
 void Mandelbrot::mousemove_event(GUI::MouseEvent& event)
 {
+    if (m_ctrl_is_pressed && m_set.m_set_type == SetType::Mandelbrot) {
```
Comment (LucasChollet) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:424]: `MouseEvent` has `ctrl()` too, no need to have `m_ctrl_is_pressed`

**Review by LucasChollet — COMMENTED** [review]
Nice to see some work on `Mandelbrot` :^)

Have you considered using the `Complex` class instead of having two separate variable like `g_c_real` and `g_c_imag`?

I haven't reviewed thoroughly the two last commits. Put them in another PR if you want the firsts to be merged quickly :)

**Comment by trflynn89** [comment]
Quick comment - this branch is 2000 commits behind the master branch.  Before opening a PR, please be sure to rebase on master and build/test your changes there.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 3 above)
Comment (nico) [inline Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: Also, I'm not sure this is a mistake. The "before" screenshots are a pretty common look of Mandelbrot viewers.

**Comment by ronak69** [comment]
> Have you considered using the Complex class instead of having two separate variable like g_c_real and g_c_imag?

Do you mean, for every complex number or just the `g_c`?

I did not use it because of efficiency (i haven't benchmarked it) and to better match with the pseudocode on [Wikipedia](https://en.wikipedia.org/wiki/Plotting_algorithms_for_the_Mandelbrot_set).

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 3 above)
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: > Also, I'm not sure this is a mistake. The "before" screenshots are a pretty common look of Mandelbrot viewers.

The linked Wikipedia article in the comment by the original author has
pseudocode for the smooth coloring rendering:

```
    while x*x + y*y ≤ (1 << 16) and iteration < max_iteration do
        xtemp:= x*x - y*y + x0
        y:= 2*x*y + y0
        x:= xtemp
        iteration:= iteration + 1

    // Used to avoid floating point issues with points inside the set.
    if iteration < max_iteration then
        // sqrt of inner term removed using log simplification rules.
        log_zn:= log(x*x + y*y) / 2
        nu:= log(log_zn / log(2)) / log(2)
        // Rearranging the potential function.
        // Dividing log_zn by log(2) instead of log(N = 1<<8)
        // because we want the entire palette to range from the
        // center to radius 2, NOT our bailout radius.
        iteration:= iteration + 1 - nu
```

Which uses `log()` instead of `sqrt()`. Also note that `nu` should be subtracted
not added.

The bands of color look (escape time algorithm) is just returning `iteration`
after the while loop (no need to calculate the fractional part).

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 2 above)
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp]: Before i figured out the "only calculate set in paint event" thing the julia set
rendering was way too slow. So this is left over from me trying to optimize
stuff. Is cast-to-int actually faster than `floor(float)`?

removed for now.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 1 above)
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp]: Just a fast path. removed for now.

**Comment by ronak69** [comment]
@LucasChollet 

Getting rid of the global variables is resulting in many getters and setters and overall a messy diff. I have removed the color and julia set commits for now.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp]: They do different things
Cast to int gives you an int and truncates, and integer math is in 80% of cases faster than floating point math, if you stay with integer math
Floor, floors (duh) and you still have a float

**Comment by LucasChollet** [comment]
> Do you mean, for every complex number or just the g_c?

For `g_c` we should use a `Complex` for sure, and in the loop if that leads to better readability and no performance issue why not.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 3 above)
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: The correct formula is this (from https://linas.org/art-gallery/escape/escape.html)
>      mu = m + frac = n + 1 - log (log  |Z(n)|) / log 2

But the wiki pseudocode divides by `log(2)` one more time for some reason.

I am not fully satisfied with the smooth coloring render right now; the radius=2 circle outline is still visible which most of the renders elsewhere don't have. I am trying to debug that.

Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: > the radius=2 circle outline is still visible which most of the renders elsewhere don't have.

Increasing the escape radius gives perfect smooth coloring. So no bug in the implementation.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp [hunk 10]
```diff
@@ -103,11 +103,11 @@ class MandelbrotSet {
             iteration++;
         }
 
-        if (iteration == max_iterations)
+        if (iteration == 1 || iteration == max_iterations)
```
[the remark this answers is not in the document]
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp:103]: If you zoom out really far then the `1 + iteration - log(lz / log(2)) / log(2)` term becomes negative because of `Z(n)`s becoming too big. This causes a crash because `hue` will not be `>= 0 && < 360`.

I do not see this problem faced by others on the internet. I guess the reason is they do not support zooming out really far.

Two more ways to fix this:
- clamp the output to `0`.
- don't allow zooming out really far

[the remark this answers is not in the document]
Comment (ronak69) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp]: I chose to clamp the output (`mu`) to `0`.

**Comment by ronak69** [comment]
https://github.com/SerenityOS/serenity/blob/c7e22c7a727b68f97ff3fc8c92b7b9256127c9ee/Userland/Demos/Mandelbrot/Mandelbrot.cpp#L120-L129

```cpp
        double hue = 360.0 * (iterations / max_iterations);
        if (hue >= 360.0)
            hue = 0.0;
```

Those two ways are equivalent (except some floating point math errors).

Should i add a commit for this change?

**Review by LucasChollet — APPROVED** [review]

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp]: That makes more sense to me, the previous solution would have returned a black color for negative mu.

**Comment by LucasChollet** [comment]
> Those two ways are equivalent (except some floating point math errors).
> 
> Should i add a commit for this change?

Yeah, sure, but let's do that in another PR.

File: Userland/Demos/Mandelbrot/Mandelbrot.cpp (hunk 3 above)
Comment (nico) [reply Userland/Demos/Mandelbrot/Mandelbrot.cpp:109]: Right, my point was that the existing look is common too, probably because it reveals a lot more structure. Thanks for updating the commit description!
